src/main/kotlin/uk/gov/dluhc/printapi/database/TwelveByteId.kt

```diff
@@ -0,0 +1,157 @@
+package uk.gov.dluhc.printapi.database
+
+import java.nio.ByteBuffer
+import java.security.SecureRandom
+import java.util.Date
+import java.util.concurrent.atomic.AtomicInteger
+
+/**
+ * A globally unique identifier, heavily based on [mongodb's ObjectId class](https://github.com/mongodb/mongo-java-driver/blob/master/bson/src/main/org/bson/types/ObjectId.java)
+ *
+ * Consists of 12 bytes, divided as follows:
+ *
+ * ```
+ *   | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 |
+ *   |---------------|-----------|-------|-------------|
+ *   | time          |rnd 1      | rnd 2 | increment   |
+ * ```
+ *
+ * Where:
+ *   * `time` is the timestamp since epoch rounded to second precision. The first 4 bytes of the integer are used.
+ *   * `rnd 1` is a random integer. This is a static member. The first 3 bytes of the integer are used.
+ *   * `rnd 2` is a random short. This is a static member. The first 2 bytes of the short are used.
+ *   * `increment` is a sequential increment, starting from a random integer. The first 3 bytes of the integer are used.
+ *
+ *  `rnd 1` and `rnd 2` are static members, meaning that all instances of [TwelveByteId] generated in the same JVM will have
+ *  the same `rnd 1` and `rnd 2` values. The net result is that bytes [4, 5, 6, 7, 8] of 2 instances generated in the same JVM
+ *  will be the same.
+ *
+ *  `increment` is an [AtomicInteger] and is a static member seeded from a random number. The use of `getAndIncrement()` means
+ *  this is a thread safe way of generating an incremental counter.
+ *
+ * An instance of [TwelveByteId] is a 12 byte identifier which serializes via it's `toString()` method into a 24 character hex string.
+ */
+class TwelveByteId {
+
+    private val timestamp: Int
+    private val counter: Int
+    private val randomValue1: Int
+    private val randomValue2: Short
+
+    companion object {
+        private const val OBJECT_ID_LENGTH = 12
+        private const val LOW_ORDER_THREE_BYTES = 0x00ffffff
+
+        // Use primitives to represent the 5-byte random value.
+        private val RANDOM_VALUE1: Int = SecureRandom().nextInt(0x01000000)
+        private val RANDOM_VALUE2: Short = SecureRandom().nextInt(0x00008000).toShort()
+
+        private val NEXT_COUNTER = AtomicInteger(SecureRandom().nextInt())
+
+        private val HEX_CHARS = charArrayOf(
+            '0', '1', '2', '3', '4', '5', '6', '7',
+            '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'
+        )
+
+        private fun dateToTimestampSeconds(date: Date): Int =
+            (date.time / 1000).toInt()
+
+        fun get(): TwelveByteId = TwelveByteId()
+    }
+
+    constructor() : this(Date())
+
+    private constructor(date: Date) : this(
```

Comment:
These private constructors should all get wrapped up into the main no-args constructors, but it was a case of copying mongo's class to get the basics working 👍